File: src/vm.c
#include <stdarg.h>
#include <stdio.h>

#include "common.h"
#include "compiler.h"
#include "debug.h"
#include "memory.h"
#include "vm.h"

VM vm;

static void reset_stack() {
    vm.stack_top = vm.stack;
    vm.frame_count = 0;
}

static void runtime_error(const char* format, ...) {
    va_list args;
    va_start(args, format);
    vfprintf(stderr, format, args);
    va_end(args);
    fputs("\n", stderr);

    CallFrame* frame = &vm.frames[vm.frame_count - 1];
    size_t instruction = frame->ip - frame->function->chunk.code - 1;
    int line = frame->function->chunk.lines[instruction];

    fprintf(stderr, "[line %d] in script\n", line);
    reset_stack();
}

void init_vm() {
    reset_stack();
    vm.objects = NULL;
    init_table(&vm.strings);
}

void free_vm() {
    free_table(&vm.strings);
    free_objects();
}

void push(Value value) {
    *vm.stack_top = value;
    vm.stack_top++;
}

Value pop() {
    vm.stack_top--;
    return *vm.stack_top;
}

static Value peek(int distance) {
    return vm.stack_top[-1 - distance];
}

static bool is_falsey(Value value) {
    return IS_NIL(value) || (IS_BOOL(value) && !AS_BOOL(value));
}

static InterpretResult run() {
    CallFrame* frame = &vm.frames[vm.frame_count - 1];

#define READ_BYTE() (*frame->ip++)
#define READ_CONSTANT() (frame->function->chunk.constants.values[READ_BYTE()])
#define READ_SHORT() (frame->ip += 2, (uint16_t)((frame->ip[-2] << 8) | frame->ip[-1]))
#define BINARY_OP(value_type, op) \
    do { \
        if (!IS_NUMBER(peek(0)) || !IS_NUMBER(peek(1))) { \
            runtime_error("Operands must be numbers."); \
            return INTERPRET_RUNTIME_ERROR; \
        } \
        int b = AS_NUMBER(pop()); \
        int a = AS_NUMBER(pop()); \
        push(value_type(a op b)); \
    } while (false)

    for (;;) {
#ifdef DEBUG_TRACE_EXECUTION
        printf("          ");
        for (Value* slot = vm.stack; slot < vm.stack_top; slot++) {
            printf("[ ");
            print_value(*slot);
            printf(" ]");
        }
        printf("\n");
        disassemble_instruction(
            &frame->function->chunk,
            (int)(frame->ip - frame->function->chunk.code),
        );
#endif

        uint8_t instruction = READ_BYTE();
        switch (instruction) {
            case OP_CONSTANT: push(READ_CONSTANT()); break;
            case OP_NIL: push(NIL_VAL); break;
            case OP_TRUE: push(BOOL_VAL(true)); break;
            case OP_FALSE: push(BOOL_VAL(false)); break;
            case OP_ADD: BINARY_OP(NUMBER_VAL, +); break;
            case OP_SUBTRACT: BINARY_OP(NUMBER_VAL, -); break;
            case OP_MULTIPLY: BINARY_OP(NUMBER_VAL, *); break;
            case OP_DIVIDE: BINARY_OP(NUMBER_VAL, /); break;
            case OP_EQUAL: {
               Value b = pop();
               Value a = pop();
               push(BOOL_VAL(values_equal(a, b)));
               break;
            }
            case OP_NOT_EQUAL: {
               Value b = pop();
               Value a = pop();
               push(BOOL_VAL(!values_equal(a, b)));
               break;
            }
            case OP_LESS: BINARY_OP(BOOL_VAL, <); break;
            case OP_GREATER: BINARY_OP(BOOL_VAL, >); break;
            case OP_LESS_EQUAL: BINARY_OP(BOOL_VAL, <=); break;
            case OP_GREATER_EQUAL: BINARY_OP(BOOL_VAL, >=); break;
            case OP_NOT: push(BOOL_VAL(is_falsey(pop()))); break;
            case OP_NEGATE:
                if (!IS_NUMBER(peek(0))) {
                    runtime_error("Operand must be a number.");
                    return INTERPRET_RUNTIME_ERROR;
                }
                push(NUMBER_VAL(-AS_NUMBER(pop())));
                break;
            case OP_POP: pop(); break;
            case OP_PUSH_DOWN: {
                Value a = pop();
                pop();
                push(a);
                break;
            }
            case OP_GET_VARIABLE: {
                uint8_t slot = READ_BYTE();
                push(vm.stack[slot]);
                break;
            }
            case OP_PRINT: print_value(pop()); printf("\n"); push(NIL_VAL); break;
            case OP_JUMP: {
                uint16_t offset = READ_SHORT();
                frame->ip += offset;
                break;
            }
            case OP_JUMP_IF_FALSE: {
                uint16_t offset = READ_SHORT();
                if (is_falsey(pop())) frame->ip += offset;
                break;
            }
            case OP_JUMP_IF_FALSE_ELSE_POP: {
                uint16_t offset = READ_SHORT();
                if (is_falsey(peek(0))) {
                    frame->ip += offset;
                } else {
                    pop();
                }
                break;
            }
            case OP_JUMP_IF_TRUE_ELSE_POP: {
                uint16_t offset = READ_SHORT();
                if (is_falsey(peek(0))) {
                    pop();
                } else {
                    frame->ip += offset;
                }
                break;
            }
            case OP_RETURN: return INTERPRET_OK;
        }
    }

#undef BINARY_OP
#undef READ_SHORT
#undef READ_CONSTANT
#undef READ_BYTE
}

InterpretResult interpret(const char* source) {
    ObjFunction* function = compile(source);
    if (function == NULL) return INTERPRET_COMPILE_ERROR;

    push(OBJ_VAL(function));
    CallFrame* frame = &vm.frames[vm.frame_count++];
    frame->function = function;
    frame->ip = function->chunk.code;
    frame->slots = vm.stack;

    return run();
}

File: src/vm.rs
use std::{
    io::{self, BufRead},
    rc::Rc,
};

use crate::{
    compiler::Op,
    value::{Function, Value},
};

pub struct VM<'src> {
    stack: Vec<Value<'src>>,
    frames: Vec<Frame<'src>>,
}

struct Frame<'src> {
    function: Rc<Function<'src>>,
    ip: u16,
    offset: u16,
}

impl<'src> VM<'src> {
    pub fn new(function: Rc<Function<'src>>) -> Self {
        VM {
            stack: Vec::new(),
            frames: vec![Frame {
                function,
                ip: 0,
                offset: 0,
            }],
        }
    }

    pub fn run(&mut self) -> Value {
        let stdin = io::stdin();
        let mut frame = self.frames.last_mut().unwrap();
        loop {
            println!("{:?}", self.stack);
            println!("{:?}", frame.function.chunk.code[frame.ip as usize]);
            // stdin.lock().lines().next().unwrap().unwrap();
            match frame.function.chunk.code[frame.ip as usize] {
                Op::Nil => self.stack.push(Value::Nil),
                Op::False => self.stack.push(Value::Bool(false)),
                Op::True => self.stack.push(Value::Bool(true)),
                Op::Constant(i) => self
                    .stack
                    .push(frame.function.chunk.constants[i as usize].clone()),
                Op::Var(i) => self
                    .stack
                    .push(self.stack[(i + frame.offset) as usize].clone()),
                Op::Jump(i) => {
                    frame.ip += i;
                    continue;
                }
                Op::JumpIfFalse(i) => {
                    if self.stack.last().unwrap().is_falsey() {
                        frame.ip += i;
                        continue;
                    }
                }
                Op::Call(args) => {
                    match self
                        .stack
                        .get(self.stack.len() - 1 - args as usize)
                        .unwrap()
                    {
                        // TODO check arity
                        Value::Function(function) => {
                            if args != function.arity {
                                todo!();
                            }
                            self.frames.push(Frame {
                                function: function.clone(),
                                ip: 0,
                                offset: self.stack.len() as u16 - args,
                            });
                            frame = self.frames.last_mut().unwrap();
                            continue;
                        }
                        Value::Native(native) => {
                            if args != native.arity {
                                todo!();
                            }
                            let value =
                                (native.f)(&self.stack[(self.stack.len() - args as usize)..]);
                            self.stack.truncate(self.stack.len() - args as usize - 1);
                            self.stack.push(value);
                        }
                        _ => todo!(),
                    }
                }
                Op::Return => {
                    {
                        let value = self.stack.pop().unwrap();
                        let frame = self.frames.pop().unwrap();
                        if self.frames.is_empty() {
                            return value;
                        }
                        self.stack.truncate(frame.offset as usize - 1);
                        self.stack.push(value);
                    }
                    frame = self.frames.last_mut().unwrap();
                }
                Op::Pop => {
                    self.stack.pop().unwrap();
                }
                Op::Squash => {
                    let value = self.stack.pop().unwrap();
                    self.stack.pop().unwrap();
                    self.stack.push(value);
                }
                Op::Unreachable => panic!("unreachable"),
            }
            frame.ip += 1;
        }
    }
}
